nemo/collections/tts/modules/fastpitch.py

```diff
@@ -261,12 +276,37 @@ def forward(
             if self.learn_alignment and pitch.shape[-1] != pitch_predicted.shape[-1]:
                 # Pitch during training is per spectrogram frame, but during inference, it should be per character
                 pitch = average_pitch(pitch.unsqueeze(1), attn_hard_dur).squeeze(1)
+            elif not self.learn_alignment:
+                # If alignment is not learnt attn_hard_dur is None, hence durs_predicted is used in average_pitch
+                pitch = average_pitch(pitch.unsqueeze(1), durs_predicted).squeeze(1)
             pitch_emb = self.pitch_emb(pitch.unsqueeze(1))
         else:
             pitch_emb = self.pitch_emb(pitch_predicted.unsqueeze(1))
 
         enc_out = enc_out + pitch_emb.transpose(1, 2)
 
+        # Predict energy
+        if self.energy_predictor is not None:
+            energy_pred = self.energy_predictor(enc_out, enc_mask).squeeze(-1)
+
+            if energy is not None:
+                # Average energy over characters
+                if self.learn_alignment:
+                    energy_tgt = average_pitch(energy.unsqueeze(1), attn_hard_dur)
+                else:
+                    energy_tgt = average_pitch(energy.unsqueeze(1), durs_predicted)
+                energy_tgt = torch.log(1.0 + energy_tgt)
```

Comment:
I think we should move the log compression into the data loader, where we do all of our other feature preprocessing/normalization. Otherwise there will be a conflict when we later refactor the pitch and energy to use the same code flow.

Reply:
Do you mean we log(energy_tgt) in the dataloader and hence before averaging? But won't that lead to mathematical inconsistencies? like log(a+b) != log(a) + log(b), I think we will run into such scenarios.

Reply:
Yes, the inconsistency is why I bring it up now to avoid backwards incompatibility.

I doubt taking the log before or after the averaging will make a difference in the final model accuracy. Which is why I suggest doing it beforehand as its more flexible and lets us add in other processing/configurations (eg. mean/var normalization), if needed in the future.